### ISSUE TITLE
Need repos for Devuan/Debian!

### ISSUE THREAD

**Issue by crhy** [FreeRDP/Remmina#977]
We support Remmina 1.2+ versions only.
Please do not report bugs occurring on older versions!

Please give us some details about the systems you are using:
* Client (OS name and version): 
* Remmina version (remmina --version): 
* Desktop environment (GNOME, Unity, KDE, ..): 
* Connecting to (OS and version): 
* Connecting via (RDP, VNC, ...): 
* RDP issues only, freerdp version (xfreerdp --version):

**Comment by crhy** [comment]
Big remmina fan, use it all day, almost every day.  But I'm migrating everything to devuan, and currently devuan is on version 1.1.0.  I'd love to be able to point to repos and get something newer, so long as it doesn't bring systemd along.

**Comment by antenore** [comment]
@giox069 @nastasi we were chatting about this on IRC and so I asked @crhy to open an issue, so that we can track the email we have got.

**Comment by dark-penguin** [comment]
I'm also using Remmina all the time, but mostly for VNC connections (Remmina kicks all the ass in the world even without RDP). And installing it manually without repos is fine by me, but installing 270 MB of development packages on each computer in order to build it is NOT.

I've just successfully built Remmina and FreeRDP on the latest Jessie following the instructions for Ubuntu 14.04, the only difference being that instead of libgnutls-dev , on Jessie it has to be libgnutls28-dev . So, it does build; now we just need a package... or at least a static binary.

- Is it possible to build Remmina (and, less importantly, FreeRDP) statically so that it doesn't depend on anything? (I can see "all build options" for FreeRDP, following the link in its README, but not for Remmina.)
- Most of those packages are probably only required for build-time. Which ones do I have to install manually if I want to build it on one machine and then install it to another one? That's got to be much less than "all of them", right?..

**Comment by giox069** [comment]
@dark-penguin, no, static build is currently not implemented, and is quite complex, because remmina depends on a lot of exernal libraries (gtk, glib, libssh, libfreerdp/libwinpr, libvnc...).
For debian packaging, you can take a look at the `debian/` subdir in the code, @nastasi did a great job. You can use the `debian/` subdir both for building your .debs or for finding remmina dependencies in debian/control file.
One day I hope we will produce a working [snapcraft](http://snapcraft.io/) package that will include all dependencies. 90% is already done, but it currently does not work.
Please be aware that the debian/ subdirectory may disappear one day... or stay forever. We have not yet decided if keeping packaging files, not only for debian, inside our upstream project is a wise choice or not :)

**Comment by antenore** [comment]
One huge dependency that you can turn off is the webkit part that it's used to compile the survey.

**Comment by dark-penguin** [comment]
> static build is currently not implemented, and is quite complex, because remmina depends on a lot of exernal libraries (gtk, glib, libssh, libfreerdp/libwinpr, libvnc...)

Indeed, those are pretty common; it wouldn't make sense.

> take a look at the debian/ subdir in the code

I'm new to packaging (more like, have no idea about it); I'll take a hack at it...
In debian/control file, I only see that all packages depend on ${shlibs:Depends} and ${misc:Depends}, which is not very informative as I couldn't find any mention of "shlibs" or "misc" anywhere else...

> We have not yet decided if keeping packaging files, not only for debian, inside our upstream project is a wise choice or not :)

The version of Remmina in Debian is 1.1.2-3 in unstable and testing, and 1.1.1-2 in stable. THAT old! It is utterly broken, and nobody notices any bug reports. So if you don't provide packages for Debian, there will be no Remmina in Debian except for the "unstable" development version for those who can build it and is ok with having a lot of development packages for it. And it's easier for you than for anyone else... so I hope you decide to make a Debian package and keep it for sake of Debian. :) I'm so sick of that fact that I'm preparing to abandon ship with Debian, but everything else seems to be as unstable as Fedora these days...

Another option would be simply providing a binary ("stable", not development) built for Debian with instructions which dependencies to install in order for it to work. That would minimize your trouble and make it at least posible to install it on Debian.

> One huge dependency that you can turn off is the webkit part that it's used to compile the survey.

That would be really helpful. And how do I do it?.. There is no list of build options for Remmina.
The thing is, I can have all my dependencies for Remmina 1.1.1-2 installed easily. But it seems that remmina-next introduces several times more dependencies, is that correct?..

**Comment by giox069** [comment]
> In debian/control file, I only see that all packages depend on ${shlibs:Depends} and ${misc:Depends}, which is not very informative as I couldn't find any mention of "shlibs" or "misc" anywhere else...

I have investigated a bit more, and yes, it seems that ${shlibs:Depends} is built at package build time. Unfortunately I'm not an expert here, so I cannot help further. I can paste the output of `apt-cache depends remmina` for the Ubuntu 16.04 remmina-next PPA package, if you like.

Please keep in mind that we are not package mantainers, and we can't spend our very limited time in packaging problems. We are doing our best effort in finding package mantainers and helping them, but currently we can't invest too much time in packaging.

> The version of Remmina in Debian is 1.1.2-3 in unstable and testing, and 1.1.1-2 in stable. THAT old! It is utterly broken, and nobody notices any bug reports.

I agree with you. Currurrently all major distributions includes Remmina 1.1 by default, because they all cannot include FreeRDP 2.0 due to its unstable state. We had a lot of collaboration from the maintainers of remmina for OpenSuse, Fedora, Ubuntu and Arch, so you can find precompiled packages for Remmina 1.2 in their "alternative" distribution systems (PPA, Copr, AUR...). Unfortunately we [had no similar help for Debian](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=835840)

In short: we would be very happy if our Debian users will be able to have precompiled Remmina 1.2 binaries, but currently we are unable to provide it, due to lack of expertise and time.
So... if someone can help us in packaging for Debian and then maintaining a .deb repository, we will be very happy.

> That would be really helpful. And how do I do it?.. There is no list of build options for Remmina.

Try to add `-DWITH_SURVEY=off` to cmake commnad line.

**Comment by antenore** [comment]
> That would be really helpful. And how do I do it?.. There is no list of build options for Remmina.
> The thing is, I can have all my dependencies for Remmina 1.1.1-2 installed easily. But it seems that remmina-next introduces several times more dependencies, is that correct?.. 

When you build remmina from the command line, you have several options and some are described in our wiki  https://github.com/FreeRDP/Remmina/wiki

During the build time, when you execute the cmake command, i.e. : 

```sh
cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_SSE2=ON -DCMAKE_INSTALL_PREFIX:PATH=/opt/remmina_devel/remmina -DCMAKE_PREFIX_PATH=/opt/remmina_devel/freerdp --build=build ..  
```
You have messages like 

```
--   Disable this using "-DWITH_GCRYPT=OFF".
--   Disable this using "-DWITH_AVAHI=OFF".
--   Disable this using "-DWITH_LIBSSH=OFF".
--   Disable this using "-DWITH_APPINDICATOR=OFF".
--   Disable this using "-DWITH_GETTEXT=OFF".
--   Disable this using "-DWITH_LIBSSH=OFF".
--   Disable this using "-DWITH_FREERDP=OFF".
--   Disable this using "-DWITH_TELEPATHY=OFF".
--   Disable this using "-DWITH_ZLIB=OFF".
--   Disable this using "-DWITH_SPICE=OFF".
--   Disable this using "-DWITH_LIBSECRET=OFF".
```
These are tricky to turn off, but you will save quite a lot of space doing it right.
As @giox069 said you can turn off the survey with -DWITH_SURVEY=off

And if you love remmina ;-) add a comment to the Debian thread mentioned by @giox069 

Hope it helps (I let you discover the right receipt for you).

**Comment by dark-penguin** [comment]
I did just before you posted this comment. :)

Debian does not have anything like PPA or AUR , so it's not that easy. The only option is to maintain your own repository. I lack expertise in this regard, but maybe that can be fixed with time...

Oh, I didn't even take care to read my output while compiling. %) Thank you! I'll try to fiddle with all this new information...

**Comment by antenore** [comment]
Thanks a lot I really appreciated it :-)

**Comment by crhy** [comment]
While we're at it, if we can get it working in Wheezy (main Debian target
still I'd guess), then we can probably get it working in Devuan, which
would be very useful to me, since I'm migrating everything to Devuan and I
use the heck out of Remmina.

Thanks for all the great work guys!

cRhyLove

**Comment by antenore** [comment]
Hey @crhy support our cause here please : 

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=835840

**Comment by dark-penguin** [comment]
@crhy What, does it not build on Devuan?.. AFAIK Devuan Jessie is basically Debian Jessie without systemd, and Remmina doesn't have anything to do with systemd, so if it should build on Devuan Jessie.

I thought that maybe we could approach the Devuan maintainers with our request to include remmina-next, which would solve both our problems, but I don't think they have any "maintainers" with that much free time on their hands yet... I will probably need to learn to package things - if I'm still on Debian by then...

I've tried building Remmina without FreeRDP. It builds just fine, so it's totally possible to package it without "unstable" FreeRDP. Also, I've tried building it WITH FreeRDP, but removing the FreeRDP directory afterwards. Again, it did work just fine. So it's also possible to package both of them, and put FreeRDP in "recommends" or "suggests" and not "depends".

I've also tried building it on Wheezy while I'm at it (without libsecret and survey, and installing libgstreamer 0.10 instead of 1.0). I could build FreeRDP with no problem, but building Remmina failed with an error I couldn't make sense of. I don't think we really need it on wheezy anyway.

Just in case for anyone who wants to build it on Wheezy, here is the error:
```
[  5%] Building C object remmina/CMakeFiles/remmina.dir/src/remmina.c.o
/home/user/Remmina/remmina/src/remmina.c: In function 'remmina_on_command_line':
/home/user/Remmina/remmina/src/remmina.c:108:2: error: unknown type name 'GVariantDict'
/home/user/Remmina/remmina/src/remmina.c:113:2: warning: implicit declaration of function 'g_application_command_line_get_options_dict' [-Wimplicit-function-declaration]
/home/user/Remmina/remmina/src/remmina.c:113:7: warning: assignment makes pointer from integer without a cast [enabled by default]
/home/user/Remmina/remmina/src/remmina.c:115:2: warning: implicit declaration of function 'g_variant_dict_lookup_value' [-Wimplicit-function-declaration]
/home/user/Remmina/remmina/src/remmina.c:128:2: warning: implicit declaration of function 'g_variant_dict_lookup' [-Wimplicit-function-declaration]
/home/user/Remmina/remmina/src/remmina.c: At top level:
/home/user/Remmina/remmina/src/remmina.c:206:58: error: unknown type name 'GVariantDict'
/home/user/Remmina/remmina/src/remmina.c: In function 'main':
/home/user/Remmina/remmina/src/remmina.c:227:2: warning: implicit declaration of function 'gdk_set_allowed_backends' [-Wimplicit-function-declaration]
/home/user/Remmina/remmina/src/remmina.c:255:2: error: 'remmina_on_local_cmdline' undeclared (first use in this function)
/home/user/Remmina/remmina/src/remmina.c:255:2: note: each undeclared identifier is reported only once for each function it appears in
/home/user/Remmina/remmina/src/remmina.c:257:2: warning: implicit declaration of function 'g_application_add_main_option_entries' [-Wimplicit-function-declaration]
make[2]: *** [remmina/CMakeFiles/remmina.dir/src/remmina.c.o] Error 1
make[1]: *** [remmina/CMakeFiles/remmina.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by dark-penguin** [comment]
I just noticed that in the compilation notes section, it says that you have to build it with -DWITH_LIBSSH=OFF as LIBSSH is too old in Wheezy. But I can't be bothered to set up another test virtual machine with Wheezy to do all that, so... if anyone wants to build it on Wheezy, try that.

**Comment by dark-penguin** [comment]
Here is the best temporary solution we can have. I've managed to build a binary distribution for Jessie - not a *.deb, but a *.tgz .

Notes:
- It is built for x86-64
- Freerdp is built without pulseaudio (does anyone use pulseaudio? it used to hang Jessie mercilessly.)
- Remmina is built without "survey" (I don't know what it is, but it saves a lot of dependencies) and without libsecret (that annoying password manager constantly asking for a password to save your password).

HOW TO INSTALL REMMINA-NEXT ON JESSIE (x86-64):
---
- Install the required libraries:
For remmina:
```
sudo apt-get install libvncclient0 libtelepathy-glib0 libappindicator3-1 libavahi-ui-gtk3-0 libgcrypt20 libgtk2.0-0 libssh-4 libvte-2.90-9 dbus-x11
```
For freerdp (skip this if you don't want freerdp):
```
sudo apt-get install libavcodec56 libgstreamer1.0-0 libgstreamer-plugins-base1.0-0
```

- Download `remmina_devel.zip` (attached). Unzip it to get `remmina.tgz` and `freerdp.tgz` separately (for some reason, Github didn't allow me to upload .tgz - it asked me to zip them first).
- Extract them into /opt/remmina_devel : 
```
sudo mkdir /opt/remmina_devel
cd /opt/remmina_devel
sudo tar xvf ~/remmina.tgz
sudo tar xvf ~/freerdp.tgz
```
(If you don't want freerdp, just don't extract it here.)

- If you want freerdp, you must register the libraries (otherwise skip this):
```
echo /opt/remmina_devel/freerdp/lib | sudo tee /etc/ld.so.conf.d/freerdp_devel.conf > /dev/null
sudo ldconfig
```
- If you want to launch remmina/freerdp by simply typing `remmina` without the full path, create the links:
```
sudo ln -s /opt/remmina_devel/remmina/bin/remmina /usr/local/bin/
sudo ln -s /opt/remmina_devel/freerdp/bin/xfreerdp /usr/local/bin/
```
- Launch remmina by typing `remmina` or `/opt/remmina_devel/remmina/bin/remmina`, if you did not want to link it in the previous step.

[remmina_devel.zip](https://github.com/FreeRDP/Remmina/files/497323/remmina_devel.zip)

**Comment by dark-penguin** [comment]
Oh yeah, you should probably remember to `apt-get autoremove --purge remmina* freerdp* ` first, though I don't know if not removing them will cause any harm...

**Comment by giox069** [comment]
@dark-penguin: thank you for the "distribution guide" for Jessie.
About Wheezy which does not compile: maybe we lost compatibility with wheezy during the last year, and there were some problems in the past (wheezy has libssh too old, GTK+ is too old too). So, I don't think remmina will work well with Wheezy. I have seen that Devuan Jessie is in beta now, @crhy are you planning to switch to Devuan Jessie ?

**Comment by antenore** [comment]
@dark-penguin this is a nice to have, thank you!

I think it deserves a pull request or at lest a wiki in Remmina and we can put in place an automatic build? 

@giox069 Just saw your answer, I agree

**Comment by dark-penguin** [comment]
@giox069 No, I think that was because I didn't care enough to find the note about disabling libssh in your instruction, or disable other options to try and do it. I don't intend to use it on Wheezy - it's even older than Jessie, which is old enough already...

@antenore I don't know what a "pull request" is yet, but do what you want with it :)
If you want to set up an automatic build, then it would probably be nice to have an i386 version as well, and you should probably consider the options more carefully, which I did not care to do because I was mostly building it for myself and not for a very wide circle of users:
- We probably shouldn't disable PulseAudio, because someone will want to use it. It won't abort on systems without PulseAudio, right? It's a "capability", not a "dependency". (At least it does work on my machine, and I don't have PulseAudio.)
- I don't use those password-protected password managers, and I think there is no other way to disable libsecret other than build without it. But maybe I'm in the minority, and most people would prefer to keep their stored passwords stored securely?..
- I still don't know what does "survey" capability mean. Is it something useful?..

**Comment by dark-penguin** [comment]
Here you are!
I've learned how to build Debian packages; I'm still totally in the dark about how to do it PROPERLY (according to the Debian policies, keeping correct versions, updating from upstream...), but at least I managed to take those Ubuntu 14.04 source packages and build binary packages that would run on Jessie.
I tried not to change anything, but in order to build it, I had to remove the debian/source directory to revert to the "old source format" (apparently, the only difference is that it would not automatically track any changes I've made to the source, which allowed me to modify debian/rules and build it with desired compile options). In theory, I should have done a huge amount of work and change everything to reflect the changes and make it into "a Debian package"; but I did not, so this remains "an Ubuntu package which works on Debian Jessie".
I've built freerdp and remmina without any changes, and then I've built freerdp-minimal (without PulseAudio and Wayland) and remmina-minimal (without survey), for both amd64 and i386. You are welcome to put it somewhere for downloading. :)
Still, unlike PPA or AUR, this is not a "well-maintained repository" from which you can install stuff automatically, leaving it up to the system to decide exactly what to download. So, I would suggest that in order to make it easier for the users, we should treat it like "Download .DEB for Jessie", and only put the necessary files for manual downloading.
[remmina.zip](https://github.com/FreeRDP/Remmina/files/504304/remmina.zip)
[remmina-minimal.zip](https://github.com/FreeRDP/Remmina/files/504306/remmina-minimal.zip)
[freerdp.zip](https://github.com/FreeRDP/Remmina/files/504307/freerdp.zip)
[freerdp-minimal.zip](https://github.com/FreeRDP/Remmina/files/504308/freerdp-minimal.zip)

**Comment by dark-penguin** [comment]
Only now I've realized some good things.

1) Compiling a "minimal" variant wasn't worth it. libpulse and libwayland are already installed on my machine, because a lot of things depend on them - though I'm certainly not using any of these. And the same goes for most dependencies of libwebkit2gtk - it only wants one more package...

2) Trying to build it without libsecret wasn't worth it either. You simply don't install the "gnome" plugin. %) I was greedily installing all plugins available, and I've even made a "todo" note to learn what "SPICE" is since I can use it with Remmina, but I didn't notice that the "gnome" plugin is a libsecret plugin and not another connection protocol. %) Now I'm completely happy about my remote desktop software. :)

**Comment by dark-penguin** [comment]
Okay, here are more proper ones. After spending a day with this, I got a better at it.
Those were build from Ubuntu 14.04 source packages without any changes other than fixing the "source format": it says "3.0 (native)", but it's in fact "3.0 (quilt)". After fixing this, it builds perfectly. And now those are generally cleaner.
Also, if you have an "automatic build system", I can probably explain to you how to build those...
[Remmina.zip](https://github.com/FreeRDP/Remmina/files/504394/Remmina.zip)
[FreeRDP.zip](https://github.com/FreeRDP/Remmina/files/504395/FreeRDP.zip)

**Comment by antenore** [comment]
That's splendid! I'm mobile now, we will discuss later. You are now a contributor ;-)

**Comment by dark-penguin** [comment]
Building this stuff is supposed to be very easy - easy enough to set up an automatic build. But there is a problem with that Ubuntu source package that makes it a little harder (requires a little workaround). I'll discuss this with PPA maintainers - either there's a good reason and I do not understand something, or they can hopefully fix this, and then building the Debian package automatically should be easy. :)

**Comment by antenore** [comment]
I completely forgot tobtell you

**Comment by antenore** [comment]
@nastasi fyi

**Comment by tomascassidy** [comment]
Devuan has some [documentation on how to package for Devuan](https://talk.devuan.org/t/os-documentation-how-to-package-for-devuan/568/14).

NB: I'm not a Devuan developer, just a daily user of Remmina/FreeRDP on Ubuntu and hoping to switch to Devuan when Remmina is working there.

**Comment by antenore** [comment]
WARNING:
The Remmina project is moving to gitlab.com, in order to keep receiving notification about this issue,
you should have a gitlab account and link it to github.
The migration will happen in the next days.
Thanks for your support and understanding.

cc @antenore @crhy @dark-penguin @giox069 @tomascassidy
